### ISSUE TITLE
pms.pid is preventing UMS from starting

### ISSUE THREAD

**Issue by robertbaker** [UniversalMediaServer/UniversalMediaServer#834]
I been having an issue ever since the 6.x betas. I'm running Windows 10 Pro x64. 

UMS will run fine on the first run, but if you quit and run it again, it will not start. I tracked it down to the pms.pid file. If I delete that file from `%programdata%\UMS\data` then UMS will run again (until I shut it down).

I'm assuming that pms.pid is a file to prevent UMS from running more than one instance. I'm a c# dev, but I think the concept should be similar. You should be able to achieve a single instance without needing a file at all, using a Mutex.

**Comment by Nadahar** [comment]
@robertbaker There are multiple issues with the current "only allow one instance" implementation, some of which is listed here: #668.

There should be better ways to do this even from Java, but there are huge limitations. Javas approach to being cross platform is heavy abstraction from everything implementation specific (and useful). This means that we don't have access to fundamental things like the Windows API or even the Windows registry without using error prone hacks. I'm not sure what kind of a Mutex you're thinking of for keeping track of another running instance, but I'm assuming it's something not available to us. Any ideas for a better implementation are welcome though, please shed some light on that Mutex implementation.

UMS also runs on Linux and OS X, so we can't have any Windows specific implementations. I've been trying to find better ways to do this, but until now I've always stumbled upon something that's not available from Java.

**Comment by robertbaker** [comment]
Here is the way to do it in c#: http://sanity-free.org/143/csharp_dotnet_single_instance_application.html

Seems like the java implementation would be to open a socket on localhost and check for that on startup. I disabled the option for now, I only run 1 instance anyway at login.

**Comment by Nadahar** [comment]
@robertbaker The socket solution sounds like a ugly hack to me, but it might be an approach worth looking in to.

**Comment by rhuneai** [comment]
I am/was having the same issue. Thank you for finding a workaround @robertbaker!
In case it helps, I am running Win 7 x64, Java 8 u73 and UMS 6.0.0 (just noticed 6.1.0 is out).
The pms.pid file is created on launch and is not removed on close. Manually deleting the file allows me to run UMS again. Running as admin doesn't fix it.
Once closed, the javaw.exe process is no longer running.
Sing out if you want any specific testing done. I'll probably grab 6.1.0 soon and see how I go.

**Comment by Sami32** [comment]
@robertbaker @rhuneai Did the latest UMS version solved this issue ?
@Nadahar did make a commit since then https://github.com/UniversalMediaServer/UniversalMediaServer/commit/b80aaffe47818e70a3530c6d0a7e0d2d6ac5581d

**Comment by Nadahar** [comment]
@Sami32 Unless something is done that I haven't noticed, this is not resolved. The reason I guess is that nobody has figured out exactly what is the right way to do it. Java has serious limitations communicating directly with the OS, which is what you'd normally do to check if it's already running. In addition, convention varies between the platforms, and the typical permission setup do as well. File locking is also different on Windows than "the others" making Windows even less suited to this way to determine if something is running. The fact that UMS doesn't run as it's own process but as the JVM makes it even harder to find a proper way. It might be that the reserved port "hack" is the best after all... but it's quite ugly imo. What if another program want to use that port?

This problem has implications for #668, #236, #489 and #739 as well.

**Comment by Sami32** [comment]
@Nadahar On Windows XP, whatever i choice "run a single instance" or not, it will only add a _javaw.exe_ process, using 4 times less memory that the first one, without any other GUI appear, but if i close the GUI these others _javaw.exe_ processes will still stay alive.

**Comment by Nadahar** [comment]
@Sami32 Yes, the problem is how to know that UMS is already running on all OS'es without admin/root permissions. I haven't been able to figure it out yet.

**Comment by Sami32** [comment]
@Nadahar Maybe it's a sutpid question...but why do we not use the _pms.pid_ file to write all the PIDs and when UMS is closed use it to close them.
Because for now, in this file i saw only one PID inside, also if i try to open 3 UMS instances; the PID change but only one value is kept in this file.

**Comment by Nadahar** [comment]
@Sami32 The whole PID approach is flawed in many ways imo. First, there's the question of permissions (read and write), second there's the question of location (all UMS instances must check the same location if there should be any point to it) and third there's the buggy Windows file locking that doesn't always release. Also, if UMS is killed by crashing or via task manager, the PID won't be updated and starting it again would be impossible - and thus the current logic is to forcefully kill the running UMS if you start a new instance. Not what I'd want if I clicked the icon by mistake, and I suspect it's the same for many others.

I did look at this earlier, but ran into trouble finding a proper way to do it. The most logical way to me would be to create a global mutex when starting that would be closed by the OS when the process ended. I got stuck in a lot of mud when trying to figure out how to do this from Java though.

http://stackoverflow.com/questions/3194227/getting-exclusive-system-wide-lock-in-java

**Comment by Sami32** [comment]
@Nadahar Thanks for trying clarify the things but to be honest it's seem to be far up of my head...

Do you mean something like that ?
https://crunchify.com/what-is-java-semaphore-and-mutex-java-concurrency-multithread-explained-with-example/

If you didn't used FileLock http://www.linuxtopia.org/online_books/programming_books/thinking_in_java/TIJ314_030.htm it's certainly because it doesn't fulfill your need.

**Comment by Nadahar** [comment]
> Do you mean something like that ?
https://crunchify.com/what-is-java-semaphore-and-mutex-java-concurrency-multithread-explained-with-example/

Yes and no. The key is **global**. If it's a mutex or a semaphore doesn't really matter. The link above is about controlling access to resources but it's on a thread level (that is, within the same process). What is needed here is some process level/OS level/global mechanism that can control shared access across different UMS instances.

> If you didn't used FileLock http://www.linuxtopia.org/online_books/programming_books/thinking_in_java/TIJ314_030.htm it's certainly because it doesn't fulfill your need.

FileLock share some of the same issues as the PID approach. There's still issues with file permissions, shared file location and the fact that a crashed or terminated instance will leave the lock in place preventing start of a new instance.
